src/model/stock_predictor_manager.py

```diff
@@ -0,0 +1,218 @@
+
+        no_improvement_count: int = 0
+        best_val_loss: float = sys.float_info.max
+        loss: torch.Tensor = torch.tensor(0)  # pylint: disable=no-member
+
+        for epoch in range(self.config.epochs):
+            for _, (inputs, targets) in enumerate(train_loader):
+                self.optimizer.zero_grad()
+                outputs: torch.Tensor = self.stock_predictor(inputs)
+                loss = self.criterion(outputs[:, -1], targets[:, -1, 0])
+                loss.backward()
+                self.optimizer.step()
+
+            val_loss: float = 0
+            with torch.no_grad():
+                for inputs, targets in val_loader:
+                    outputs = self.stock_predictor(inputs)
+                    val_loss += self.criterion(outputs[:, -1], targets[:, -1, 0])
+            val_loss /= len(val_loader)
+
+            if verbose:
+                print(
+                    f"Epoch {epoch + 1}: Train Loss = {loss:.6f}, Validation Loss = {val_loss:.6f}"
+                )
+
+            if val_loss + self.config.min_delta >= best_val_loss:
+                no_improvement_count += 1
+                if no_improvement_count == self.config.patience:
+                    if verbose:
+                        print(
+                            f"No improvement in validation loss after {self.config.patience} \
+                                epochs. Stopping training..."
+                        )
+                    break
+            else:
+                best_val_loss = val_loss
+                no_improvement_count = 0
+
+    def predict_values(self, days: int) -> pd.DataFrame:
+        """Predict stock value for next x days"""
+
+        # Gets newest seq_length dates from the data source.
+        data: pd.DataFrame = self._get_newest_dates()
+
+        # Get first date for which the stock value will be predicted.
+        current_date: datetime.date = pd.to_datetime(data.index[0]).date() + datetime.timedelta(
+            days=1
+        )
+
+        date_to_predict: datetime.date = pd.to_datetime(data.index[0]).date() + datetime.timedelta(
+            days=days
+        )
+
+        date_index: pd.DatetimeIndex = pd.date_range(
+            start=current_date, end=date_to_predict, freq="D"
+        )
+        predicted_values: pd.DataFrame = pd.DataFrame(index=date_index, columns=["value"])
+
+        # Predict values until date reaches date_to_predict.
+        while current_date <= date_to_predict:
+            # Scale data and convert it to tensor
+            input_seq_scaled: np.ndarray = self.scaler.fit_transform(data)
+            input_seq_tensor: torch.Tensor = (
+                torch.tensor(input_seq_scaled).float().unsqueeze(0)  # pylint: disable=no-member
+            )
+
+            # Predict output for the next date
+            with torch.no_grad():
+                predicted_output_tensor: torch.Tensor = self.stock_predictor(input_seq_tensor)
```

Comment:
Czy nie brakuje self.stock_predictor.eval() ?
- przejście z "trybu" uczenia na "tryb" normalny

Reply:
I forgot to add this earlier, now it has been added.